User branch

BJSherman80 opened the pull request:
When a user visits the root path they should be on the welcome page which includes:
  - [x] Welcome message
  - [x] Brief description of the application
  - [x] Button to Log in
  - [x] Link to Registration
When a user visits the '/registration' path they should see a form to register.
The form should include:
  - [x] Email
  - [x] Password
  - [x]  Password Confirmation
  - [x]  Register Button

BJSherman80 committed:
create test for welcome page w/ buttons and login form

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
add routes for welcome page 

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create welcome view w/ form 

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create view template for registration form w/ no info yet

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create welcome controller w/ index action

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create user controller w/ new action

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
add bcrypt/ shoulda_matchers/orderly gems

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
add shoulda matchers method 

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create user model with validations and create duplicte email method 

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
add flash message into the application body

Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
add gem lock file 

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create helper methods and current_user

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create sessions controller w create action

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create show/create actions and private params 

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create form for new user registration


Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create view for user show 

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create routes for user registration and sessions

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create migrations and schema for users

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
create test for user login w all sad path
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
Create validations test for user 

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

Co-Authored-By: Jake Heft <60531761+jakeheft@users.noreply.github.com>

BJSherman80 committed:
Create user registration tests


Co-authored-by: Jake Heft <60531761+jakeheft@users.noreply.github.com>

dcoleman21 commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,15 @@
+    redirect_to '/dashboard'
+  else
+    flash[:error] = "Sorry, your credentials are bad."
+    redirect_to '/'
```
We can change line 11 to say `redirect_to root_path

dcoleman21 commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,34 @@
+  end
+
+  def show
+    @user = User.find(session[:user_id])
```
we might think about adding a sad path here if a user is not signed in. For example: render file: '/public/401' unless current_user

jakeheft replied:
Good idea Dani; I'll add to refactor wishlist

dcoleman21 commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,34 @@
+      flash[:notice] = user.errors.full_messages.uniq
+      render :new
+    elsif params[:password] != params[:password_confirmation]
+      flash[:failure] = "Password and Password Confirmation fields did not match."
```
Maybe we just say 'Passwords do not match'

dcoleman21 commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,4 @@
+<h1>Welcome, <%= @user.name %>! </h1>
```
I think that this should be current_user.name because they couldn't get to this page unless they are a user. 

Then we could get rid of the if statement and just have the button to logout ( <%= button_to 'Logout', '/logout', method: :delete %>

dcoleman21 commented on config/routes.rb:
```diff
@@ -1,3 +1,9 @@
+  post '/registration', to: 'users#create'
+  get '/dashboard', to: 'users#show'
+  #session
+   post '/login', to: 'sessions#create'
```
then when they logout, we can add a  delete '/logout', to: 'sessions#destroy'

dcoleman21 commented on spec/features/Users/login_spec.rb:
```diff
@@ -0,0 +1,50 @@
+
+    expect(page).to have_content("Sorry, your credentials are bad.")
+  end
+
```
maybe we also add a test here that the user can logout and a sad path test for displaying an error message if they try to visit the dashboard page when not logged in (ie 401/not authorized to access page)

dcoleman21 reviewed (changes requested):
Hey team, I have left some comments in here. Have a look and then we can make adjustments tomorrow.

jakeheft committed:
Add destroy method to sessions controller, refactor using path helper

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Brett Sherman 
<64512380+BJSherman80@users.noreply.github.com>

jakeheft committed:
Add method require_user, commented out for now

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Brett Sherman 
<64512380+BJSherman80@users.noreply.github.com>

jakeheft committed:
Refactor show action for sad path

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Brett Sherman 
<64512380+BJSherman80@users.noreply.github.com>

jakeheft committed:
Update logout function to DRY code

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Brett Sherman 
<64512380+BJSherman80@users.noreply.github.com>

jakeheft committed:
Add logout path to destroy session

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Brett Sherman 
<64512380+BJSherman80@users.noreply.github.com>

jakeheft committed:
Add 401 page for sad path test when not logged in

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Brett Sherman 
<64512380+BJSherman80@users.noreply.github.com>

jakeheft committed:
Refactor to Logout and change to path helper

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Brett Sherman 
<64512380+BJSherman80@users.noreply.github.com>

jakeheft committed:
Create log out test

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Brett Sherman 
<64512380+BJSherman80@users.noreply.github.com>

jakeheft committed:
Change to logout

Co-authored-by: Dani Coleman 
<60626984+dcoleman21@users.noreply.github.com>
Co-authored-by: Brett Sherman 
<64512380+BJSherman80@users.noreply.github.com>
